rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+the worker and not the request to the artifact service.  If the origin is an
+IPv4 or IPv6, a mapping of the address and IP will occur to find the backing
+storage.  If the origin is not an IPv4 or IPv6 address and is identical to a
+set of known and mappable identifiers (e.g. `s3_us-west-2`), then that will be
+used to map the token to find the backing storage.
+
+### Deletetion
+
+```
+DELETE /artifacts/:name
+DELETE /caches/artifacts/:name
+```
+
+Delete all copies of the named Artifact if using first endpoint.  The second
+endpoint would attempt delete all of the cached copies in the various
+locations.  This cache purging must return a `202 Accepted` response and not
+`200/204` due to the nature of how caches work in the real world.
+
+### Retreival
+
+```
+GET /artifacts/:name[?max_time=30]
+```
+
+This endpoint returns a 302 redirect to the location of the Artifact which is
+closest to the request's IP address.  If needed, this will initiate from the
+original backing store into a different storage location.  It will wait up to a
+configurable amount of time before redirecting to the non-optimal original
+backing store.
+
+In order to support running in environments where requests are limted in time,
+the `attempts` query string option will specify roughly how many seconds the
+client is willing to wait for an artifact.  The default value should be based
+on the size of the artifact.  For example, it will wait 2s for each 100MB of
+artifact size.  While waiting, the service will issue intermediate redirects
+to itself to reduce the waiting time.
+
+Example inside heroku:
+```
+10:00:00 <-- GET /artifacts/my_file?max_time=90
+10:00:25 --> 302 /artifacts/my_file?max_time=65
+10:00:25 <-- GET /artifacts/my_file?max_time=65
+10:00:50 --> 302 /artifacts/my_file?max_time=40
+10:00:50 <-- GET /artifacts/my_file?max_time=40
+10:00:55 --> 302 http://us-west-2.s3.amazonaws/artifacts/my_file
+```
+
+### Redirects and Errors
+```
+GET/PUT/DELETE /redirects/:name
+GET/PUT/DELETE /errors/:name
+```
+
+Since redirects aren't artifacts and errors aren't artifacts, they will have
+their own endpoints.  There's no reason to mix them into the same tables and
+logic for the full artifacts, so we will have endpoints for their creation and
+management.
+
+For the `PUT` method of redirects, the body will look like:
+
+```javascript
+{
+  url: 'https://www.google.com'
+}
+```
+
+The `GET` method will redirect to the URL value from the `PUT` body
+
+For the `PUT` method of errors, the body will look like:
+
+```javascript
+{
+  reason: 'error code',
+  message: 'error message'
+}
+```
+
+The `GET` method will return the metadata with a `403 Forbidden` response code.
+
+
+# Open Questions
+
+* Will we move old artifacts over?
+* Will we move deprecated artifact types (s3, azure) over?
+* How will we support data centers where we have no native object store
+* What auth scheme will we use between Queue and Artifacts
+* Will we use Content Addressable storage?
```

Comment:
It would.  I also think this is something we don't need to figure out right now, since the mapping of name that the queue would use is unrelated to the internal storage name.

Let's decide to not decide on this point until we're further along. maybe

Reply:
Can we make it an open question how to deduplicate private artifacts...